### PR TITLE
API-2532: replaced compound id with three individual ids in response p…

### PR DESCRIPTION
**Pull request by paulblakehmrc**
…ayload and removed SubscriptionIdentifier along with FieldDefinitionIdentifier

### CODE REVIEW

File: app/uk/gov/hmrc/apisubscriptionfields/controller/FieldsDefinitionController.scala [hunk 1]
```diff
@@ -33,12 +33,12 @@ class FieldsDefinitionController @Inject() (service: FieldsDefinitionService) ex
   import JsonFormatters._
 
   private def notFoundResponse(rawApiContext: String, rawApiVersion: String) =
-    NotFound(JsErrorResponse(ErrorCode.NOT_FOUND_CODE, s"Id ($rawApiContext, $rawApiVersion) was not found"))
+    NotFound(JsErrorResponse(ErrorCode.NOT_FOUND_CODE, s"Fields definition ($rawApiContext, $rawApiVersion) was not found"))
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/controller/FieldsDefinitionController.scala]: I would update the error message as:
```
s"Fields definition not found for ($rawApiContext, $rawApiVersion)"
```

File: app/uk/gov/hmrc/apisubscriptionfields/controller/SubscriptionFieldsController.scala [hunk 2]
```diff
@@ -39,7 +39,7 @@ class SubscriptionFieldsController @Inject()(service: SubscriptionFieldsService)
   }
 
   private def notFoundMessage(rawClientId: String, rawApiContext: String, rawApiVersion: String): String = {
-    s"Id ($rawClientId, $rawApiContext, $rawApiVersion) was not found"
+    s"Subscription field ($rawClientId, $rawApiContext, $rawApiVersion) was not found"
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/controller/SubscriptionFieldsController.scala]: I would update it as:
```
s"Subscription fields not found for ($rawClientId, $rawApiContext, $rawApiVersion)"
```

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 3]
```diff
@@ -93,26 +93,22 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
...
-  override def fetch(identifier: SubscriptionIdentifier): Future[Option[SubscriptionFields]] = {
-    val selector = selectorForIdentifier(identifier)
+  override def fetch(clientId: ClientId, apiContext: ApiContext, apiVersion: ApiVersion): Future[Option[SubscriptionFields]] = {
+    val selector = selectorForSubscriptionField(clientId.value, apiContext.value, apiVersion.value)
     Logger.debug(s"[fetch] selector: $selector")
     collection.find(selector).one[SubscriptionFields]
   }
 
-  private def selectorForIdentifier(clientId: String, apiContext: String, apiVersion: String): JsObject = {
+  private def selectorForSubscriptionField(clientId: String, apiContext: String, apiVersion: String): JsObject = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Could you rename it as `selectorForSubscriptionFields` please?

File: test/acceptance/AcceptanceTestSpec.scala [hunk 4]
```diff
@@ -40,7 +40,7 @@ trait AcceptanceTestSpec extends FeatureSpec
   protected val ValidRequest = FakeRequest()
     .withHeaders(RequestHeaders.ACCEPT_HMRC_JSON_HEADER)
 
-  protected def idEndpoint(clientId: String, apiContext: String, apiVersion: String) =
+  protected def subscriptionEndpoint(clientId: String, apiContext: String, apiVersion: String) =
```
Comment (gokyo) [inline test/acceptance/AcceptanceTestSpec.scala]: could you rename it as `subscriptionFieldsEndpoint` please?

File: test/acceptance/AcceptanceTestSpec.scala [hunk 5]
```diff
@@ -40,7 +40,7 @@ trait AcceptanceTestSpec extends FeatureSpec
...
     s"/field/application/$clientId/context/$apiContext/version/$apiVersion"
 
   protected def clientIdEndpoint(clientId: String) = s"/field/application/$clientId"
```
Comment (gokyo) [inline test/acceptance/AcceptanceTestSpec.scala]: could you rename it as `byClientIdEndpoint ` please?

File: test/acceptance/ApiSubscriptionFieldsSpec.scala [hunk 6]
```diff
@@ -98,13 +97,13 @@ class ApiSubscriptionFieldsSpec extends AcceptanceTestSpec
       val sfr = contentAsJson(resultFuture).validate[SubscriptionFieldsResponse]
 
       sfr.isSuccess shouldBe true
-      sfr.get should matchPattern { case SubscriptionFieldsResponse(FakeRawIdentifier, _, SampleFields1) => }
+      sfr.get should matchPattern { case SubscriptionFieldsResponse(`fakeRawClientId`, "acontext", "1.0.2", _, SampleFields1) => }
     }
 
-    scenario("the API is called to GET with a known fields identifier") {
+    scenario("the API is called to GET with a known fieldId") {
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsSpec.scala]: fieldsId

File: test/acceptance/ApiSubscriptionFieldsSpec.scala [hunk 7]
```diff
@@ -98,13 +97,13 @@ class ApiSubscriptionFieldsSpec extends AcceptanceTestSpec
...
 
-      Given("a request with a known identifier")
-      val requestId = ValidRequest.copyFakeRequest(method = GET, uri = idEndpoint(fakeRawClientId, fakeRawContext, fakeRawVersion))
+      Given("a request with a known fieldId")
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsSpec.scala]: fieldsId

File: test/unit/uk/gov/hmrc/apisubscriptionfields/service/SubscriptionFieldsServiceSpec.scala [hunk 8]
```diff
@@ -29,71 +29,69 @@ class SubscriptionFieldsServiceSpec extends UnitSpec with SubscriptionFieldsTest
...
-        SubscriptionFieldsResponse(id = "TODO: remove this field", fieldsId = SubscriptionFieldsId(subscriptionFields1.fieldsId), fields = subscriptionFields1.fields),
-        SubscriptionFieldsResponse(id = "TODO: remove this field", fieldsId = SubscriptionFieldsId(subscriptionFields2.fieldsId), fields = subscriptionFields2.fields)
+        SubscriptionFieldsResponse(clientId = subscriptionFields1.clientId, apiVersion = subscriptionFields1.apiVersion, apiContext = subscriptionFields1.apiContext, fieldsId = SubscriptionFieldsId(subscriptionFields1.fieldsId), fields = subscriptionFields1.fields),
+        SubscriptionFieldsResponse(clientId = subscriptionFields2.clientId, apiVersion = subscriptionFields2.apiVersion, apiContext = subscriptionFields2.apiContext, fieldsId = SubscriptionFieldsId(subscriptionFields2.fieldsId), fields = subscriptionFields2.fields)
       )))
     }
 
-    "return None when no entry exist in the repo when get by composite id is called" in {
-      (mockSubscriptionFieldsIdRepository fetch _) expects FakeSubscriptionIdentifier returns None
+    "return None when no entry exists in the repo when get by subscription field is called" in {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apisubscriptionfields/service/SubscriptionFieldsServiceSpec.scala]: `return None when no entry exists in the repo`

File: test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala [hunk 9]
```diff
@@ -50,34 +50,34 @@ class FieldsDefinitionServiceSpec extends UnitSpec with FieldsDefinitionTestData
...
 
-      val result = await(service.get(FakeFieldsDefinitionIdentifier))
+      val result = await(service.get(FakeContext, FakeVersion))
 
       result shouldBe None
     }
 
-    "return Some when entry exists in the repo when get by identifier is called" in {
-      (mockFieldsDefinitionRepository fetch (_: FieldsDefinitionIdentifier)) expects FakeFieldsDefinitionIdentifier returns Some(FakeFieldsDefinition)
+    "return Some when entry exists in the repo when get by fields definition is called" in {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala]: `return Some when definition exists in the repo`

File: test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala [hunk 10]
```diff
@@ -50,34 +50,34 @@ class FieldsDefinitionServiceSpec extends UnitSpec with FieldsDefinitionTestData
       result shouldBe FieldsDefinitionResponse(Seq(FakeFieldDefinitionUrl, FakeFieldDefinitionString))
     }
 
-    "return None when no entry exist in the repo when get by identifier is called" in {
-      (mockFieldsDefinitionRepository fetch (_: FieldsDefinitionIdentifier)) expects FakeFieldsDefinitionIdentifier returns None
+    "return None when no entry exist in the repo when get by fields definition is called" in {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala]: `return None when no definition exists in the repo`

File: test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala [hunk 11]
```diff
@@ -124,33 +124,33 @@ class SubscriptionFieldsRepositorySpec extends UnitSpec
...
     "return an empty list when clientId is not found" in {
-      await(repository.fetchByClientId("APP_ID_DOES_NOT_EXIST_IN_DB")) shouldBe List()
+      await(repository.fetchByClientId("CLIENT_ID_DOES_NOT_EXIST_IN_DB")) shouldBe List()
     }
 
   }
 
-  "fetch using compound Id" should {
-    "retrieve the correct record for a compound Id" in {
+  "fetch using subscription field" should {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala]: `"fetch using clientId, apiContext, apiVersion"`

File: test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala [hunk 12]
```diff
@@ -124,33 +124,33 @@ class SubscriptionFieldsRepositorySpec extends UnitSpec
...
+    "retrieve the correct record for a subscription field" in {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala]: "retrieve the correct record"

File: test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala [hunk 13]
```diff
@@ -124,33 +124,33 @@ class SubscriptionFieldsRepositorySpec extends UnitSpec
...
       val apiSubscription = createApiSubscriptionFields()
       await(repository.save(apiSubscription)) shouldBe true
       collectionSize shouldBe 1
 
-      await(repository.fetch(FakeSubscriptionIdentifier)) shouldBe Some(apiSubscription)
+      await(repository.fetch(FakeClientId, FakeContext, FakeVersion)) shouldBe Some(apiSubscription)
     }
 
-    "return `None` when the `id` doesn't match any record in the collection" in {
+    "return `None` when the subscription field doesn't match any record in the collection" in {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala]: `"return `None` when no subscription fields are found in the collection"`

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala [hunk 14]
```diff
@@ -70,9 +70,9 @@ class ApiSubscriptionFieldsUnhappySpec extends AcceptanceTestSpec
       contentAsJson(resultFuture) shouldBe JsErrorResponse(NOT_FOUND_CODE, s"FieldsId (${FakeRawFieldsId.toString}) was not found")
     }
 
-    scenario("the API is called to GET an unknown application identifier") {
+    scenario("the API is called to GET an unknown client identifier") {
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: replace `client identifier` with `application clientId`

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala [hunk 15]
```diff
@@ -70,9 +70,9 @@ class ApiSubscriptionFieldsUnhappySpec extends AcceptanceTestSpec
...
 
-      Given("the API is called to GET an unknown application identifier")
+      Given("the API is called to GET an unknown client identifier")
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: replace `client identifier` with `application clientId`

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala [hunk 16]
```diff
@@ -34,10 +34,10 @@ class ApiSubscriptionFieldsUnhappySpec extends AcceptanceTestSpec
   feature("Subscription-Fields") {
     Logger.logger.info(s"App.mode = ${app.mode.toString}")
 
-    scenario("the API is called to GET an unknown subscription identifier") {
+    scenario("the API is called to GET an unknown subscription field") {
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: `the API is called to GET non-existing subscription fields`

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala [hunk 17]
```diff
@@ -34,10 +34,10 @@ class ApiSubscriptionFieldsUnhappySpec extends AcceptanceTestSpec
...
 
-      Given("the API is called to GET an unknown subscription identifier")
-      val request = ValidRequest.copyFakeRequest(method = GET, uri = idEndpoint(fakeRawClientId, fakeRawContext, fakeRawVersion))
+      Given("the API is called to GET an unknown subscription field")
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: `the API is called to GET non-existing subscription fields`

File: test/acceptance/ApiSubscriptionFieldsSpec.scala [hunk 18]
```diff
@@ -175,14 +174,13 @@ class ApiSubscriptionFieldsSpec extends AcceptanceTestSpec
...
-
+      sfr.get should matchPattern { case SubscriptionFieldsResponse(`fakeRawClientId`, "acontext", "1.0.2", _, SampleFields2) => }
     }
 
-    scenario("the API is called to DELETE a known subscription identifier") {
+    scenario("the API is called to DELETE a known subscription field") {
 
-      Given("a request with a known identifier")
-      val request = ValidRequest.copyFakeRequest(method = DELETE, uri = idEndpoint(fakeRawClientId, fakeRawContext, fakeRawVersion))
+      Given("a request with a known subscription field")
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsSpec.scala]: `a request with existing subscription fields`

File: test/acceptance/ApiSubscriptionFieldsSpec.scala [hunk 19]
```diff
@@ -175,14 +174,13 @@ class ApiSubscriptionFieldsSpec extends AcceptanceTestSpec
       val sfr = contentAsJson(resultFuture).validate[SubscriptionFieldsResponse]
 
       sfr.isSuccess shouldBe true
-      sfr.get should matchPattern { case SubscriptionFieldsResponse(FakeRawIdentifier, _, SampleFields2) => }
-
+      sfr.get should matchPattern { case SubscriptionFieldsResponse(`fakeRawClientId`, "acontext", "1.0.2", _, SampleFields2) => }
     }
 
-    scenario("the API is called to DELETE a known subscription identifier") {
+    scenario("the API is called to DELETE a known subscription field") {
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsSpec.scala]: `the API is called to DELETE existing subscription fields`

File: test/acceptance/ApiSubscriptionFieldsSpec.scala [hunk 20]
```diff
@@ -154,13 +153,13 @@ class ApiSubscriptionFieldsSpec extends AcceptanceTestSpec
       val sfr = contentAsJson(resultFuture).validate[BulkSubscriptionFieldsResponse]
 
       sfr.isSuccess shouldBe true
-      sfr.get should matchPattern { case BulkSubscriptionFieldsResponse(Seq(SubscriptionFieldsResponse(FakeRawIdentifier, _, SampleFields1))) => }
+      sfr.get should matchPattern { case BulkSubscriptionFieldsResponse(Seq(SubscriptionFieldsResponse(`fakeRawClientId`, "acontext", "1.0.2", _, SampleFields1))) => }
     }
 
-    scenario("the API is called to store more values for an existing subscription identifier") {
+    scenario("the API is called to store more values for an existing subscription field") {
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsSpec.scala]: `the API is called to update existing subscription fields`

File: test/acceptance/ApiSubscriptionFieldsSpec.scala [hunk 21]
```diff
@@ -133,12 +132,12 @@ class ApiSubscriptionFieldsSpec extends AcceptanceTestSpec
       val sfrFieldsId = contentAsJson(resultFuture).validate[SubscriptionFieldsResponse]
 
       sfrFieldsId.isSuccess shouldBe true
-      sfrFieldsId.get should matchPattern { case SubscriptionFieldsResponse(FakeRawIdentifier, `fieldsId`, SampleFields1) => }
+      sfrFieldsId.get should matchPattern { case SubscriptionFieldsResponse(`fakeRawClientId`, "acontext", "1.0.2", `fieldsId`, SampleFields1) => }
     }
 
-    scenario("the API is called to GET with a known application identifier") {
+    scenario("the API is called to GET with a known client identifier") {
```
Comment (gokyo) [inline test/acceptance/ApiSubscriptionFieldsSpec.scala]: `the API is called to GET existing subscription fields by application clientId`

File: app/uk/gov/hmrc/apisubscriptionfields/controller/FieldsDefinitionController.scala (hunk 1 above)
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/apisubscriptionfields/controller/FieldsDefinitionController.scala]: Done

File: app/uk/gov/hmrc/apisubscriptionfields/controller/SubscriptionFieldsController.scala (hunk 2 above)
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/apisubscriptionfields/controller/SubscriptionFieldsController.scala]: Done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 3 above)
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Done

File: test/acceptance/AcceptanceTestSpec.scala (hunk 4 above)
Comment (paulblakehmrc) [reply test/acceptance/AcceptanceTestSpec.scala]: Done

File: test/acceptance/AcceptanceTestSpec.scala (hunk 5 above)
Comment (paulblakehmrc) [reply test/acceptance/AcceptanceTestSpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsSpec.scala (hunk 6 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsSpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsSpec.scala (hunk 7 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsSpec.scala]: Done

File: test/unit/uk/gov/hmrc/apisubscriptionfields/service/SubscriptionFieldsServiceSpec.scala (hunk 8 above)
Comment (paulblakehmrc) [reply test/unit/uk/gov/hmrc/apisubscriptionfields/service/SubscriptionFieldsServiceSpec.scala]: Done

File: test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala (hunk 9 above)
Comment (paulblakehmrc) [reply test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala]: Done

File: test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala (hunk 10 above)
Comment (paulblakehmrc) [reply test/unit/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionServiceSpec.scala]: Done

File: test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala (hunk 11 above)
Comment (paulblakehmrc) [reply test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala]: Done

File: test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala (hunk 12 above)
Comment (paulblakehmrc) [reply test/unit/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepositorySpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala (hunk 14 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala (hunk 15 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala (hunk 16 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala (hunk 17 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsUnhappySpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsSpec.scala (hunk 18 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsSpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsSpec.scala (hunk 19 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsSpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsSpec.scala (hunk 20 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsSpec.scala]: Done

File: test/acceptance/ApiSubscriptionFieldsSpec.scala (hunk 21 above)
Comment (paulblakehmrc) [reply test/acceptance/ApiSubscriptionFieldsSpec.scala]: Done

**Pull request merged**